src/sas/sascalc/poresize/maxEnt_method.py

```diff
@@ -0,0 +1,390 @@
+        for iter in range(IterMax):
+            ox = -2 * z / sigma                                    
+            
+            cgrad = np.dot(ox, Gqr)  # cgrad[i] = del(C)/del(f[i]), SB eq. 8
+            sgrad = -np.log(f/first_bins) / (blank*math.exp (1.0)) # sgrad[i] = del(S)/del(f[i])
+            snorm = math.sqrt(sum(f * sgrad*sgrad))                # entropy, SB eq. 22
+            cnorm = math.sqrt(sum(f * cgrad*cgrad))                # Chi^2, SB eq. 22
+            tnorm = sum(f * sgrad * cgrad)                         # norm of gradient
+            
+            a = 1.0
+            b = 1.0 / cnorm
+            if iter == 0:
+                test = 0.0     # mismatch between entropy and ChiSquared gradients
+            else:
+                test = math.sqrt( ( 1.0 - tnorm/(snorm*cnorm) )/2 ) # SB eq. 37?
+                a = 0.5 / (snorm * test)
+                b *= 0.5 / test
+            xi[0] = f * cgrad / cnorm
+            xi[1] = f * (a * sgrad - b * cgrad)
+            
+            eta[0] = np.dot(xi[0], Gqr.transpose());          # solution --> data
+            eta[1] = np.dot(xi[1], Gqr.transpose());          # solution --> data
+            ox = eta[1] / (sigma * sigma)
+            xi[2] = np.dot(ox, Gqr);          # data --> solution
+            a = 1.0 / math.sqrt(sum(f * xi[2]*xi[2]))
+            xi[2] = f * xi[2] * a
+            eta[2] = np.dot(xi[2], Gqr.transpose())           # solution --> data
+
+            # prepare the search directions for the conjugate gradient technique
+            c1 = xi.dot(cgrad) / chisq                          # C_mu, SB eq. 24
+            s1 = xi.dot(sgrad)                                  # S_mu, SB eq. 24
+
+            
+            for k in range(SEARCH_DIRECTIONS):
+                for l in range(k+1):
+                    c2[k][l] = 2 * sum(eta[k] * eta[l] / sigma/sigma) / chisq
+                    s2[k][l] = -sum(xi[k] * xi[l] / f) / blank
+
+            # reflect across the body diagonal
+            for k, l in ((0,1), (0,2), (1,2)):
+                c2[k][l] = c2[l][k]                     #  M_(mu,nu)
+                s2[k][l] = s2[l][k]                     #  g_(mu,nu)
+    
+            beta[0] = -0.5 * c1[0] / c2[0][0]
+            beta[1] = 0.0
+            beta[2] = 0.0
+            if (iter > 0):
+                w, chtarg, loop, a_new, fx, beta = self.MaxEntMove(fSum, blank, chisq, chizer, c1, c2, s1, s2)
+                
+            f_old = f.copy()                # preserve the last solution
+            f += xi.transpose().dot(beta)   # move the solution towards the solution, SB eq. 25
+        
+            # As mentioned at the top of p.119,
+            # need to protect against stray negative values.
+            # In this case, set them to RESET_STRAYS * base[i]
+            #f = f.clip(RESET_STRAYS * blank, f.max())
+            for i in range(n):
+                if f[i] <= 0.0:
+                    f[i] = RESET_STRAYS * first_bins[i]
```

Comment:
Likely faster, but untested: `f = np.array([x if x > 0.0 else RESET_STRAYS * y for x, y in zip(f, first_bins)])`

Reply:
Perhaps, but the point is to keep that code as is (after losing additions at the top of the file). It is essentially taken from GSAS code.  It is not pretty by but it works and is tested. To fully understand it so that we can rewrite would take too much time. The goal would be to replace this module in the future.